sql/schemachanger: implemented ALTER PRIMARY KEY for vanilla case

Xiang-Gu opened the pull request:
The PR implements `ALTER PRIMARY KEY` under the declarative schema
changer framework that handles the simplest, "vanilla" case like
```
CREATE TABLE t (i INT PRIMARY KEY, j INT NOT NULL)
ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (j)
```

This is the first of a series PRs where followup PRs will expand its
capabilities to be able to handle more complex cases, including
  1. Allow the requested new primary key to be hash-sharded;
  2. Consider the case where altering primary key requires us to
     modify existing secondary indexes(see the legacy schema change
     about in what cases we should rewrite)
  3. Consider the case where the old primary index is on the implicitly
     created `rowid` column, in which case we also need to drop that
     column;
  5. Consider partitioning and locality (I'm not sure what they are,
     and why they play a role when `ALTER PRIMARY KEY` but I've seen
     them in the old schema changer, so I assume we ought to do
     something about them too here).
  6. Support `ALTER PRIMARY KEY` with concurrent schema change
      statements. E.g.
 ```ALTER TABLE t ADD COLUMN k INT NOT NULL DEFAULT 30, ALTER PRIMARY KEY USING COLUMNS (j);```

related: #83932
Release note: None

Xiang-Gu committed:
sql/schemachanger: implemented `ALTER PRIMARY KEY` for vanilla case

The PR implements `ALTER PRIMARY KEY` under the declarative schema
changer framework that handles the simplest, "vanilla" case like
```
CREATE TABLE t (i INT PRIMARY KEY, j INT NOT NULL)
ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (j)
```

This is the first of a series PRs where followup PRs will expand its
capabilities to be able to handle more complex cases, including
  1. Allow the requested new primary key to be hash-sharded;
  2. Consider the case where the old primary index is on the implicitly
     created `rowid` column, in which case we also need to drop that
     column;
  3. Consider the case where altering primary key requires us to
     modify existing secondary indexes(see the legacy schema change
     about in what cases we should rewrite)
  4. Consider partitioning and locality (I'm not sure what they are,
     and why they play a role when `ALTER PRIMARY KEY` but I've seen
     them in the old schema changer, so I assume we ought to do
     something about them too here).

Release note: None

Xiang-Gu committed:
sql/schemachanger: added tests for the `ALTER PRIMARY KEY` vanilla case

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84514)
<!-- Reviewable:end -->

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,384 @@
+	"github.com/cockroachdb/errors"
+)
+
+// The implementation of `ALTER PRIMARY KEY` will be broken into multiple PRs:
```
This plan belongs in an issue and not in the code.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,384 @@
+// 2. Consider the case where the new primary key is requested to be sharded.
+// 3. Consider the case where the old primary index is on the implicitly created `rowid` column,
+//    in which case we also need to drop that column;
+// 4. Consider the case where altering primary key requires us to modify existing secondary indexes
```
This step deserves to come before the rowid changes we don't currently support.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,384 @@
+	oldPrimaryIndexElem := mustRetrievePrimaryIndexElement(tableElems, tn.String())
+	oldPrimaryIndexNameElem := mustRetrievePrimaryIndexNameElem(tableElems, oldPrimaryIndexElem.IndexID, tn.String())
+
+	// Resolve and drop elements from the old primary index
```
How does this compose with adding or dropping columns? Let's look at this together tomorrow afternoon. In the meantime, can I encourage you to read the DROP COLUMN PR

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84514)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r2-31:-N7RsuHq4_L1V-XtpBG3:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea59185f55713b697ebd59a604336d09ca80f563/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L31)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This plan belongs in an issue and not in the code. 
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r2-36:-N7RsS7P6jXP_NeJlkFV:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea59185f55713b697ebd59a604336d09ca80f563/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L36)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This step deserves to come before the rowid changes we don't currently support. 
</blockquote></details>

ack
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r2-60:-N7RsSvRDNoQtBd6Le2I:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea59185f55713b697ebd59a604336d09ca80f563/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L60)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

How does this compose with adding or dropping columns? Let's look at this together tomorrow afternoon. In the meantime, can I encourage you to read the DROP COLUMN PR 
</blockquote></details>

ack


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -230,6 +230,30 @@ func (b *builderState) NextViewIndexID(view *scpb.View) (ret catid.IndexID) {
+		if constraintName.ConstraintID >= ret {
+			ret = constraintName.ConstraintID + 1
+		}
+	})
```
Drive-by reviewing: you should be able to walk through all the constraint IDs in the table by doing a little rel-magic:
```
	nextID := catid.ConstraintID(1)
	b.QueryByID(table.TableID).ForEachElementStatus(func(_ scpb.Status, _ scpb.TargetStatus, e scpb.Element) {
		v, _ := screl.Schema.GetAttribute(screl.ConstraintID, e)
		if id, ok := v.(catid.ConstraintID); ok && id >= nextID {
			nextID = id + 1
		}		
	})
```
This should allow you to remove this method entirely, which would be nice.

Xiang-Gu replied:
Thanks! We have similar nextIndexID, nextColumnID method and I like having a nextConstraintID method as well (since it might be used more than once). But I do like and will change the implementation to what you suggested.

Xiang-Gu replied:
edit: in fact this method is never used in this PR (but will be in the next followup PR) so I delete it for now.

Xiang-Gu committed:
sql/catalog/tabledesc: fixed an issue for estimating number of keys for a row in an index

We encounter an issue for this case
```
CREATE TABLE t (i INT PRIMARY KEY, j INT NOT NULL, FAMILY (i), FAMILY (j));
INSERT INTO t VALUES (23, 24);
ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (j);
```
where the last `ALTER PRIMARY KEY` will result in duplicate entries
in the new primary index, because the number of estimated keys for a
row in this new primary index is incorrectly estimated to be 1, rather
than 2 (the number of column families).

ajwerner commented on pkg/sql/sem/tree/select.go:
```diff
@@ -743,6 +744,18 @@ func (d Direction) String() string {
 	return directionName[d]
 }
 
+// ToIndexColumnDirection converts tree.Direction to catpb.IndexColumn_Direction.
+func (d Direction) ToIndexColumnDirection() catpb.IndexColumn_Direction {
+	switch d {
+	case DefaultDirection, Ascending:
+		return catpb.IndexColumn_ASC
+	case Descending:
+		return catpb.IndexColumn_DESC
+	default:
+		panic(fmt.Sprintf("invalid direction %s", d))
+	}
+}
+
```
I don't like this dependency. Can we put this somewhere else? I would love if we could fully separate the catalog protobufs from the ast where neither know about each other. It'll take some time, but this is moving backwards. Currently tree doesn't import any of the protobufs.

ajwerner replied:
I'd expect the below rule to prevent this from passing. 

```
    disallowed_prefixes = [
        "pkg/sql/catalog",	        "pkg/sql/catalog",
    ],	    ],
```

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+	}
+
+	// TODO (xiang): This section contains all fall-back cases and need to
+	// 							 be removed to fully support `ALTER PRIMARY KEY`.
```
nit: bad indentation

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+	oldPrimaryIndexElem, newPrimaryIndexElem := getPrimaryIndexes(b, tbl.TableID)
+	if newPrimaryIndexElem != nil {
+		// TODO (Xiang): some other DDL stmt preceded this `ALTER PRIMARY KEY` and
+		//               thus a new primary index has already been created. We'd like
+		//               to support this use case one day
+		//               (e.g. `ALTER TABLE t ADD COLUMN ..., ALTER PRIMARY KEY ...;`).
+		//               Note that such scenarios should be caught above in
+		//               `fallBackIfConcurrentSchemaChange` and an unimplemented error
+		//               should be returned, so, here we panic with an programming error.
```
nit you don't need to wrap this like this, I think it's harder to read and annoying to maintain.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+		//               `fallBackIfConcurrentSchemaChange` and an unimplemented error
+		//               should be returned, so, here we panic with an programming error.
+		panic("programming error: new primary index has already existed.")
+	} else {
```
nit, unnest this, you don't need it nested given the panic.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+
+		colCurrentStatus, _, colElem := scpb.FindColumn(colElems)
+		if colElem == nil {
+			panic(fmt.Sprintf("programming error: resolving column %v does not give a "+
```
I'd rather you used an `errors.AssertionFailedf` error because it'll capture the stack trace. Maybe we need a helper `panicf` which combines `panic` and `errors.AssertionFailedWithDepthf`

ajwerner replied:
The other advantage of the error constructors is they will properly handle redaction management for arguemnts. If you don't construct an error and instead format a string, the higher levels will redact the entire message.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+	oldPrimaryIndexElem := mustRetrievePrimaryIndexElement(b, tbl.TableID)
+	oldPrimaryIndexKeyColumns := mustRetrieveKeyIndexColumns(b, tbl.TableID, oldPrimaryIndexElem.IndexID)
+
+	// check whether they have the same number of key columns.
+	if len(oldPrimaryIndexKeyColumns) != len(t.Columns) {
+		return false
+	}
+
+	// check whether they are both sharded or both not sharded.
+	if (oldPrimaryIndexElem.Sharding == nil) != (t.Sharded == nil) {
+		return false
+	}
+
+	// check whether all key columns (ID and directions) are the same.
```
nit: Use capital letters to start sentences.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+// there are any other concurrent schema change on this table. This is determined
+// by searching for any element that is in ABSENT (resp. PUBLIC) and targeting
+// PUBLIC (resp. ABSENT).
+func fallBackIfConcurrentSchemaChange(b BuildCtx, tableID catid.DescID) {
```
I don't think this is good enough. Consider the case where we've had a previous statement in the same transaction but all of the elements in the statement phase have moved away from their initial status. Consider just checking if the elements are not in their terminal state.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+	scpb.ForEachPrimaryIndex(b.QueryByID(tableID), func(
+		current scpb.Status, target scpb.TargetStatus, e *scpb.PrimaryIndex,
+	) {
+		// TODO (xiang): for now, we assume there is one primary index, which
+		//               will not be true when there are DDL stmts (e.g.
+		//               ADD/DROP COLUMN) before this `ALTER PRIMARY KEY`.
```
same wrapping nit

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+	return columnType
+}
+
+func mustRetrieveIndexElement(
```
One day before too long I'm going to come back and rework all of this to use rel and to be efficient. So many linear lookup loops.

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+	}
+	return indexColumns
+}
+func dropOldPrimaryIndex(b BuildCtx, tableID catid.DescID, oldPrimaryIndexID catid.IndexID) {
+	oldPrimaryIndexName := mustRetrievePrimaryIndexNameElem(b, tableID, oldPrimaryIndexID).Name
+
+	b.ResolveIndex(tableID, tree.Name(oldPrimaryIndexName), ResolveParams{
+		IsExistenceOptional: false,
+		RequiredPrivilege:   privilege.CREATE,
+	}).ForEachElementStatus(func(_ scpb.Status, target scpb.TargetStatus, e scpb.Element) {
+		if target != scpb.ToAbsent {
+			b.Drop(e)
+		}
+	})
+}
```
Is this not handled by `createNewPrimaryIndex`?

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go:
```diff
@@ -0,0 +1,797 @@
+		}
+	}
+
+	// STORED columns (empty)
```
what's this about?

ajwerner commented on pkg/sql/schema_changer_test.go:
```diff
@@ -2646,7 +2646,10 @@ CREATE TABLE t.test (k INT NOT NULL, v INT, v2 INT NOT NULL)`); err != nil {
 		// Alter the primary key of the table.
 		go func() {
-			if _, err := sqlDB.Exec(`ALTER TABLE t.test ALTER PRIMARY KEY USING COLUMNS (v2)`); err != nil {
+			if _, err := sqlDB.Exec(`
+SET use_declarative_schema_changer = off;
+ALTER TABLE t.test ALTER PRIMARY KEY USING COLUMNS (v2);
+SET use_declarative_schema_changer = on;`); err != nil {
```
move this change into the previous commit or change this to be the commit which enables the schema change. We don't want commits where tests fail. It makes bisecting things later hard.

ajwerner commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -111,7 +111,7 @@ func (desc *wrapper) GetParentSchemaID() descpb.ID {
 // IndexKeysPerRow implements the TableDescriptor interface.
 func (desc *wrapper) IndexKeysPerRow(idx catalog.Index) int {
-	if desc.PrimaryIndex.ID == idx.GetID() {
+	if desc.PrimaryIndex.ID == idx.GetID() || idx.GetEncodingType() == descpb.PrimaryIndexEncoding {
```
move this commit to be first

ajwerner reviewed (commented):
Getting close!

Xiang-Gu committed:
turn ALTER PRIMARY KEY on

A few tests failed because they test/reply on specific behavior in the
old schema changer, so we disabled them in this pr.

Xiang-Gu reviewed (commented):
@ajwerner  I've addressed your comments and it's RFAL again.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-:-N8AwJvc0h7eycLEKqKw:bgkf29f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/catalog/tabledesc/structured.go` line 114 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N800ePWGn9tALjQCYBN-r13-114:-N89egqW0sGtFwT3ux1H:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3df35006eb28e07e7e59cc5dff7651eb18643b6/pkg/sql/catalog/tabledesc/structured.go#L114)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

move this commit to be first
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 51 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-51:-N89PSFvAsg-n3hHQG2T:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L51)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: bad indentation
</blockquote></details>

done



_Code quote:_
```Go
	// TODO (xiang): This section contains all fall-back cases and need to
	// 							 be removed to fully support `ALTER PRIMARY KEY`.
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 67 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-67:-N89PhDX72V2EPoNBu-l:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L67)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit you don't need to wrap this like this, I think it's harder to read and annoying to maintain. 
</blockquote></details>

done



_Code quote:_
```Go
Note that such scenarios should be caught above in
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 69 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-69:-N89PiAT0j3fkLMLqsWf:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L69)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit, unnest this, you don't need it nested given the panic. 
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 173 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-173:-N89SmKA-ihjq_EAXOet:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L173)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The other advantage of the error constructors is they will properly handle redaction management for arguemnts. If you don't construct an error and instead format a string, the higher levels will redact the entire message. 
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 221 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-221:-N89TlSp4EUAVpwB5EAT:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L221)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: Use capital letters to start sentences.
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 253 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-253:-N89U_DxBCxqjw8HiIO0:bgf9375) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L253)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think this is good enough. Consider the case where we've had a previous statement in the same transaction but all of the elements in the statement phase have moved away from their initial status. Consider just checking if the elements are not in their terminal state. 
</blockquote></details>

you're right. Done!
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 340 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-340:-N89UlXzEXcFmM1j0AtN:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L340)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

same wrapping nit
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 408 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-408:-N89UmiCC3-YVP4KRFws:b-hzpvs4) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L408)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

One day before too long I'm going to come back and rework all of this to use rel and to be efficient. So many linear lookup loops. 
</blockquote></details>

totally supported! Thanks in advance!
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 458 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-458:-N89V7m69VutNCu1d9Iy:bx5m3wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L458)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is this not handled by `createNewPrimaryIndex`?
</blockquote></details>

you're right! This function is never used bc I forgot to delete it after I refactored the code to reuse `createNewPrimaryIndex`
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 614 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N72t6tMDlMjeKpcnuCb-r11-614:-N89VI1P8BEwihFu3XGj:b-gud6c1) ([raw file](https://github.com/cockroachdb/cockroach/blob/154055946fffa82a3a3138b40748c326265e50c6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L614)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

what's this about?
</blockquote></details>

I just want to leave a comment there to say there is no STORED columns in this case. I've removed this line.
___
*[`pkg/sql/schema_changer_test.go` line 2652 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N80fbblPJnHvpFzNGai-r12-2652:-N89eaPa596xDUwUEo2x:b-zheyq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a09ef7e8d386b40e24665b38f3cdc2bb541f34f/pkg/sql/schema_changer_test.go#L2652)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

move this change into the previous commit or change this to be the commit which enables the schema change. We don't want commits where tests fail. It makes bisecting things later hard. 
</blockquote></details>

done. I've moved it to the last commit where I turn on `ALTER PRIMARY KEY`
___
*[`pkg/sql/sem/tree/select.go` line 758 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N85MiUoRyx-jh3H-4Jz-r10-758:-N89YLPl68Sa39NaOH84:bwl2xzj) ([raw file](https://github.com/cockroachdb/cockroach/blob/bce6ced309769b3654721448c3161b5edc0bbfab/pkg/sql/sem/tree/select.go#L758)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'd expect the below rule to prevent this from passing. 

```
    disallowed_prefixes = [
        "pkg/sql/catalog",	        "pkg/sql/catalog",
    ],	    ],
```
</blockquote></details>

I've moved this utility function to be in package `scbuildstmt` as one of many helpers there.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
The `rttanalysis` failures are unexpected. You've got a couple more tests to fix. 

> Reviewed 2 of 6 files at r15, 1 of 2 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-:-N8PLrU5BqCntAl_xpb8:bqaa6dg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 128 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N8PLU8AE8muaD3HT4oi:-N8PLU8B0pgrAXQaoHOh:b-dqogpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa512cbaf6fddd8886f06efcd452a868fda4a6a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L128)):*
> ```Go
> 	b.EvalCtx().ClientNoticeSender.BufferClientNotice(b,
> 		pgnotice.Newf(
> 			"primary key changes are finalized asynchronously; "+
> ```

this is not true. If a test expects it, let's get rid of that test. 


<!-- Sent from Reviewable.io -->

Xiang-Gu committed:
sql: removed outdated notice message for ALTER PRIMARY KEY

Previously, we issued a notice to customer for `ALTER PRIMARY KEY`,
saying the altering primary key is done asynchronously and subsequent
schema change might be rejected. This was no longer true (see the slace
conversation
https://cockroachlabs.slack.com/archives/C04U1BTF8/p1640124375172500 for
some history.

This PR removes this notice and modified the test that tests this
notice.

Xiang-Gu reviewed (commented):
I've fixed the CI and address the comments. RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-:-N8j-NZ29U4SzM7b8xyZ:b-8869d2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 128 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N8PLU8AE8muaD3HT4oi:-N8j-M55EziKJaPSg7tm:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4aa512cbaf6fddd8886f06efcd452a868fda4a6a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L128)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this is not true. If a test expects it, let's get rid of that test. 
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 8 files at r7, 1 of 15 files at r10, 2 of 11 files at r16, 1 of 12 files at r18, 2 of 2 files at r19, 3 of 4 files at r20, 3 of 20 files at r21, 2 of 2 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-:-N8xoKwIARKEir9KAq3h:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_primary_key` line 7 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N8xndDiEZiGz4_FGcso:-N8xndDiEZiGz4_FGcsp:b-tbs58t) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdb3764d7a5649b54a0a0a9bbc9d313637380dea/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L7)):*
> ```
> INSERT INTO t VALUES (1, 2, 3, 4), (5, 6, 7, 8)
> 
> query T noticetrace
> ```

nit: change this to `statement ok`?



_Code quote:_
```
query T noticetrace
ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (y, z)
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 75 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N8xo1dSELrdm4lvBV0l:-N8xo1dSELrdm4lvBV0m:b34flgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdb3764d7a5649b54a0a0a9bbc9d313637380dea/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L75)):*
> ```Go
> 
> 		// Get all KEY columns from t.Columns
> 		allColumnsName2IDMapping := getAllColumnsName2IDMapping(b, tbl.TableID)
> ```

nit: I don't think the use of `2` is clearer than `To`, can you rename all these?


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84514)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/logictest/testdata/logic_test/alter_primary_key` line 7 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N8xndDiEZiGz4_FGcso:-N8y8n9_EMmaN3p0niS-:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdb3764d7a5649b54a0a0a9bbc9d313637380dea/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L7)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: change this to `statement ok`?
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 75 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/84514#-N8xo1dSELrdm4lvBV0l:-N8y91CH5Yv3dgqqgbJo:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdb3764d7a5649b54a0a0a9bbc9d313637380dea/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L75)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: I don't think the use of `2` is clearer than `To`, can you rename all these?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

postamar commented:
The merge failed due to a bunch of flakes, which ought to be fixed in the current bors batch. Retrying.

bors r+
